File: be/src/main/java/sg/flow/services/TransactionHistoryServices/TransactionHistoryServiceImpl.java
package sg.flow.services.TransactionHistoryServices;

import java.time.LocalDate;

import org.springframework.stereotype.Service;

import jakarta.websocket.SendResult;
import lombok.RequiredArgsConstructor;
import sg.flow.models.transaction.TransactionHistoryDetail;
import sg.flow.models.transaction.TransactionHistoryList;
import sg.flow.models.transfer.TransferRecepient;
import sg.flow.models.transfer.TransferRequestBody;
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository;
import sg.flow.services.BankQueryServices.DBSQueryService;

@Service
@RequiredArgsConstructor
public class TransactionHistoryServiceImpl implements TransactionHistoryService {

    private final TransactionHistoryRepository transactionHistoryRepository;
    private final DBSQueryService bankQueryService;

    @Override
    public TransactionHistoryList getLast30DaysHistoryList(int userId) {
        return transactionHistoryRepository.findTransactionBetweenDates(userId, LocalDate.now().minusDays(30),
                LocalDate.now());
    }

    @Override
    public TransactionHistoryList getMonthlyTransaction(int userId, int year, int month) {
        LocalDate currentMonthFirstDay = LocalDate.of(year, month, 1);
        LocalDate currentMonthLastDay = LocalDate.of(year, month, currentMonthFirstDay.lengthOfMonth());

        TransactionHistoryList monthlyHistoryList = transactionHistoryRepository.findTransactionBetweenDates(userId,
                currentMonthFirstDay, currentMonthLastDay);

        return monthlyHistoryList;
    }

    @Override
    public TransactionHistoryList getDailyTransaction(int userId, LocalDate date) {
        TransactionHistoryList dailyHistoryList = transactionHistoryRepository.findTransactionBetweenDates(userId, date,
                date);
        // if (date.isBefore(currentDate)) {
        // return dailyHistoryList;
        // }

        // LocalDateTime lastUpdated =
        // transactionHistoryRepository.getLastUpdatedDate();
        // TransactionHistoryList additionalHistoryList =
        // bankQueryService.getTransactionHistory(userId, );
        // dailyHistoryList.add(additionalHistoryList);
        return dailyHistoryList;
    }

    @Override
    public TransactionHistoryDetail getTransactionDetails(int userId, String transaction_id) {
        return transactionHistoryRepository.findTransactionDetailById(Long.parseLong(transaction_id));
    }

    @Override
    public TransferRecepient getRelevantRecepient(String keyword) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'getRelevantRecepient'");
    }

    @Override
    public SendResult sendTransaction(String recepientId, Double amount) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'sendTransaction'");
    }

    @Override
    public SendResult sendTransaction(TransferRequestBody sendRequestBody) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'sendTransaction'");
    }

    @Override
    public TransactionHistoryList getTransactionWithinRange(int userId, LocalDate startDate, LocalDate endDate) {
        return transactionHistoryRepository.findTransactionBetweenDates(userId, startDate, endDate);
    }

}

File: be/src/main/kotlin/sg/flow/services/TransactionHistoryServices/TransactionHistoryServiceImpl.kt
package sg.flow.services.TransactionHistoryServices

import java.time.LocalDate
import org.springframework.stereotype.Service
import sg.flow.entities.RecurringSpendingMonthly
import sg.flow.models.transaction.TransactionHistoryDetail
import sg.flow.models.transaction.TransactionHistoryList
import sg.flow.models.transfer.TransferRecepient
import sg.flow.models.transfer.TransferRequestBody
import sg.flow.repositories.recurring.RecurringSpendingRepository
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository

@Service
class TransactionHistoryServiceImpl(
        private val transactionHistoryRepository: TransactionHistoryRepository,
        private val recurringSpendingRepository: RecurringSpendingRepository
) : TransactionHistoryService {

    override suspend fun getMonthlyTransaction(
            userId: Int,
            year: Int,
            month: Int
    ): TransactionHistoryList {
        val startDate = LocalDate.of(year, month, 1)
        val endDate = startDate.withDayOfMonth(startDate.lengthOfMonth())
        return transactionHistoryRepository.findTransactionBetweenDates(userId, startDate, endDate)
    }

    override suspend fun getDailyTransaction(userId: Int, date: LocalDate): TransactionHistoryList {
        return transactionHistoryRepository.findTransactionBetweenDates(userId, date, date)
    }

    override suspend fun getTransactionDetails(
            userId: Int,
            transactionId: String
    ): TransactionHistoryDetail {
        transactionHistoryRepository.findTransactionDetailById(id = transactionId.toLong())?.let {
                detail ->
            if (detail.account?.owner?.id != userId) {
                throw IllegalArgumentException("Transaction does not belong to user")
            }
            return detail
        }
        throw IllegalArgumentException("Transaction not found")
    }

    override suspend fun getRelevantRecepient(keyword: String): TransferRecepient {
        // TODO: Implement getRelevantRecepient
        throw NotImplementedError("getRelevantRecepient not implemented")
    }

    override suspend fun getLast30DaysHistoryList(userId: Int): TransactionHistoryList {
        val endDate = LocalDate.now()
        val startDate = endDate.minusDays(30)
        return transactionHistoryRepository.findTransactionBetweenDates(
                userId,
                startDate,
                endDate,
                30
        )
    }

    override suspend fun getTransactionWithinRange(
            userId: Int,
            startDate: LocalDate,
            endDate: LocalDate
    ): TransactionHistoryList {
        return transactionHistoryRepository.findTransactionBetweenDates(userId, startDate, endDate)
    }

    override suspend fun getProcessedTransactionsForTransactionIds(userId: Int, transactionIds: List<String>): TransactionHistoryList {
        return transactionHistoryRepository.findProcessedTransactionsFromTransactionIds(userId, transactionIds)
    }

    override suspend fun getRecurringTransactionAnalysisResult(userId: Int): List<RecurringSpendingMonthly> {
        return recurringSpendingRepository.findRecurringTransactionsForUserId(userId)
    }

    override suspend fun setTransactionCategory(
        userId: Int,
        transactionId: String,
        category: String
    ): Boolean {
        val result = transactionHistoryRepository.setTransactionCategory(userId, transactionId, category);
        return result
    }

    override suspend fun setTransactionInclusion(
        userId: Int,
        transactionId: String,
        includeInSpendingOrIncome: Boolean
    ): Boolean {
        val result = transactionHistoryRepository.setTransactionInclusion(userId, transactionId, includeInSpendingOrIncome)
        return result
    }

    override suspend fun getTransactionsForAccount(
        userId: Int,
        accountNumber: String,
        bankId: String,
        oldestTransactionId: String,
        limit: Int
    ): TransactionHistoryList {
        return transactionHistoryRepository.findTransactionForAccountOlderThan(userId, bankId, accountNumber, oldestTransactionId, limit)
    }
}
